In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV
from matplotlib.colors import ListedColormap

In [2]:
# Importing the dataset
dataset = pd.read_csv('dataset.csv') #831x23
#X = dataset.iloc[:, 4:22].values
#y = dataset.iloc[:, 22].values

In [3]:
len(dataset)

831

In [4]:
.8 * 831

664.8000000000001

In [5]:
# Splitting the dataset into the Training set and Test set
#from sklearn.cross_validation import train_test_split
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

X_train = dataset.iloc[0:664, 4:22].values
y_train = dataset.iloc[0:664, 22].values
X_test = dataset.iloc[664:831, 4:22].values
y_test = dataset.iloc[664:831, 22].values

In [6]:
X_train[0]

array([2.50000000e-01, 7.36749117e-01, 7.39045764e-01, 2.25405405e+01,
       1.45961538e+01, 3.26855124e-02, 5.06329114e-02, 1.66067233e+01,
       1.07872257e+01, 7.50000000e-01, 9.47144075e-01, 9.21632653e-01,
       3.26764706e+01, 1.94655172e+01, 2.89855072e-02, 4.73469388e-02,
       3.09493255e+01, 1.79400563e+01])

In [7]:
y_train[0]

0

In [8]:
len(y_train)

664

In [9]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


## Random Forest

In [10]:
classifier = RandomForestClassifier(n_estimators =10 , criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [11]:
y_pred = classifier.predict(X_test)

In [12]:
cm = confusion_matrix(y_test, y_pred)

In [13]:
cm

array([[21, 38],
       [47, 61]])

In [14]:
classifier.score(X_test,y_test)

0.49101796407185627

In [15]:
y_pred = classifier.predict(X_train)

In [16]:
cm = confusion_matrix(y_train, y_pred)

In [17]:
cm

array([[261,   3],
       [  5, 395]])

In [18]:
n_estimators = [2,4,6,8,10]
min_samples_split = [2,3,4,5,6]
l = [2,3,4,5,6]
criterion = ['entropy','gini']
#param_grid = dict(n_estimators=n_estimators,min_samples_split=min_samples_split,criterion=criterion)

#model = RandomForestClassifier(random_state = 0)
best_test_score=0
best_train_score=0
best_n_estimators=0
min_samples_split=0
criterion=''
diff=100
for n_estimator in [2,4,6,8,10]:
    for split in [2,3,4,5,6]:
        for c in ['entropy','gini']:
            classifier =RandomForestClassifier(n_estimators=n_estimator,min_samples_split=split,criterion=c,random_state = 0)
            classifier.fit(X_train, y_train)
            #y_pred = classifier.predict(X_train)
            train_score = classifier.score(X_train,y_train)
            test_score = classifier.score(X_test,y_test)
            #print("n_estimators = {}, min_samples_split = {}, criterion = {} ,train_score = {},test_score = {}").format(n_estimator,split,c,train_score,test_score)
            #print()
            if(test_score > best_test_score):
                best_test_score=test_score
                best_train_score=train_score
                best_n_estimators=n_estimator
                best_min_samples_split=split
                best_criterion=c
                #diff = (abs(test_score-test_score))
                

print("best_test_score = {},best_train_score = {},best_n_estimators = {} ,min_samples_split = {}, criterion = {}".format(best_test_score,best_train_score,best_n_estimators,best_min_samples_split,best_criterion))        


best_test_score = 0.5688622754491018,best_train_score = 0.9743975903614458,best_n_estimators = 10 ,min_samples_split = 5, criterion = entropy


# logistic regression

In [19]:
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=0, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [20]:
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
cm

array([[14, 45],
       [13, 95]])

In [21]:
y_pred = classifier.predict(X_train)
cm = confusion_matrix(y_train, y_pred)
cm

array([[ 70, 194],
       [ 55, 345]])

In [22]:

#solver = ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
#penalty = ['l1', 'l2', 'elasticnet', 'none']
#C = [0.0001,0.001,0.01,0.1,1,10]
#max_iter = [80,90,100]
best_test_score=0
best_train_score=0
best_solver = ''
best_penalty = ''
best_C = 0
best_max_iter = 0
for solver in ['lbfgs', 'liblinear', 'sag', 'saga']:
    for penalty in ['l2']:
        for C in [0.012,0.011,0.0115,0.0116,0.0117,0.0118,0.0119,0.0121,0.0122,0.0123,0.0124]:
            for max_iter in [1,2,3]:
                
                classifier = LogisticRegression(solver=solver,penalty=penalty,C = C,max_iter=max_iter,random_state = 0)
                classifier.fit(X_train, y_train)
                
                train_score = classifier.score(X_train,y_train)
                test_score = classifier.score(X_test,y_test)
                
                if(test_score > best_test_score):
                    best_test_score = test_score
                    best_train_score = train_score
                    best_solver = solver
                    best_penalty = penalty
                    best_C = C
                    best_max_iter = max_iter
                
print("best_test_score = {},best_train_score = {},best_solver = {} ,best_penalty = {}, best_C = {}, best_max_iter = {}".format(best_test_score,best_train_score,best_solver,best_penalty,best_C,best_max_iter))        

                
                

/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.


best_test_score = 0.6706586826347305,best_train_score = 0.6099397590361446,best_solver = sag ,best_penalty = l2, best_C = 0.012, best_max_iter = 1


/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/home/amor/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached w

### best_test_score = 0.6706586826347305,best_train_score = 0.6099397590361446,best_solver = sag ,best_penalty = l2, best_C = 0.012, best_max_iter = 1

# Gaussian NB

In [23]:
classifier = GaussianNB()
classifier.fit(X_train, y_train)

train_score = classifier.score(X_train,y_train)
test_score = classifier.score(X_test,y_test)

print("train_score = {} , test_score = {}".format(train_score,test_score))


train_score = 0.6189759036144579 , test_score = 0.5688622754491018


In [24]:
cm = confusion_matrix(y_test, classifier.predict(X_test))
cm

array([[13, 46],
       [26, 82]])

In [25]:
(13.0 + 82)/(13 + 46+26+82)

0.5688622754491018

### train_score = 0.6189759036144579 , test_score = 0.5688622754491018

# SVC

In [26]:
# C = [0.0001,0.001,0.01,0.1,1,10]
# kernel = ['linear', 'poly', 'rbf', 'sigmoid', 'precomputed']
# gamma = ['scale' , 'auto']
# max_iter
best_test_score=0
best_train_score=0
best_C = 0
best_kernel = ''

for C in [0.188,0.189,0.19]:
    for kernel in ['linear', 'poly', 'rbf', 'sigmoid']:
        
        classifier = SVC(C=C,kernel=kernel,random_state = 0)
        classifier.fit(X_train, y_train)
        
        train_score = classifier.score(X_train,y_train)
        test_score = classifier.score(X_test,y_test)

        if(test_score > best_test_score):
            best_C = C
            best_test_score = test_score
            best_train_score = train_score
            best_kernel = kernel
            
print("best_test_score = {},best_train_score = {},best_C = {}, best_kernel = {}".format(best_test_score,best_train_score,best_C,best_kernel))        


best_test_score = 0.6706586826347305,best_train_score = 0.5783132530120482,best_C = 0.189, best_kernel = sigmoid


### best_test_score = 0.6706586826347305,best_train_score = 0.5783132530120482,best_C = 0.189, best_kernel = sigmoid